# Importing Library

In [646]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [647]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
plt.style.use('dark_background')

# Importing Dataset

In [648]:
# dir = "\weatherAUS.csv"
rainfall_data = pd.read_csv('/content/drive/MyDrive/MASTERS IN GEOSPATIAL TECHNOLOGIES/Master GeoTech - Course Materials and links/SJL005 - UJI - Artificial Intelligence and Machine Learning (2022 2023)/009 - My Application Work/001- Raw data/Australia Rainfall - Kaggle.zip')
rainfall_data.head(10)


Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0  2008-12-01   Albury     13.4     22.9       0.6          NaN       NaN   
1  2008-12-02   Albury      7.4     25.1       0.0          NaN       NaN   
2  2008-12-03   Albury     12.9     25.7       0.0          NaN       NaN   
3  2008-12-04   Albury      9.2     28.0       0.0          NaN       NaN   
4  2008-12-05   Albury     17.5     32.3       1.0          NaN       NaN   
5  2008-12-06   Albury     14.6     29.7       0.2          NaN       NaN   
6  2008-12-07   Albury     14.3     25.0       0.0          NaN       NaN   
7  2008-12-08   Albury      7.7     26.7       0.0          NaN       NaN   
8  2008-12-09   Albury      9.7     31.9       0.0          NaN       NaN   
9  2008-12-10   Albury     13.1     30.1       1.4          NaN       NaN   

  WindGustDir  WindGustSpeed WindDir9am  ... Humidity9am  Humidity3pm  \
0           W           44.0          W  ...        71.0         22.0   
1         WNW           44.0        NNW  ...        44.0         25.0   
2         WSW           46.0          W  ...        38.0         30.0   
3          NE           24.0         SE  ...        45.0         16.0   
4           W           41.0        ENE  ...        82.0         33.0   
5         WNW           56.0          W  ...        55.0         23.0   
6           W           50.0         SW  ...        49.0         19.0   
7           W           35.0        SSE  ...        48.0         19.0   
8         NNW           80.0         SE  ...        42.0          9.0   
9           W           28.0          S  ...        58.0         27.0   

   Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday  \
0       1007.7       1007.1       8.0       NaN     16.9     21.8         No   
1       1010.6       1007.8       NaN       NaN     17.2     24.3         No   
2       1007.6       1008.7       NaN       2.0     21.0     23.2         No   
3       1017.6       1012.8       NaN       NaN     18.1     26.5         No   
4       1010.8       1006.0       7.0       8.0     17.8     29.7         No   
5       1009.2       1005.4       NaN       NaN     20.6     28.9         No   
6       1009.6       1008.2       1.0       NaN     18.1     24.6         No   
7       1013.4       1010.1       NaN       NaN     16.3     25.5         No   
8       1008.9       1003.6       NaN       NaN     18.3     30.2         No   
9       1007.0       1005.7       NaN       NaN     20.1     28.2        Yes   

   RainTomorrow  
0            No  
1            No  
2            No  
3            No  
4            No  
5            No  
6            No  
7            No  
8           Yes  
9            No  

[10 rows x 23 columns]

In [649]:
# Description of my numeric dataset
rainfall_data.describe()

MinTemp        MaxTemp       Rainfall   Evaporation  \
count  143975.000000  144199.000000  142199.000000  82670.000000   
mean       12.194034      23.221348       2.360918      5.468232   
std         6.398495       7.119049       8.478060      4.193704   
min        -8.500000      -4.800000       0.000000      0.000000   
25%         7.600000      17.900000       0.000000      2.600000   
50%        12.000000      22.600000       0.000000      4.800000   
75%        16.900000      28.200000       0.800000      7.400000   
max        33.900000      48.100000     371.000000    145.000000   

           Sunshine  WindGustSpeed   WindSpeed9am   WindSpeed3pm  \
count  75625.000000  135197.000000  143693.000000  142398.000000   
mean       7.611178      40.035230      14.043426      18.662657   
std        3.785483      13.607062       8.915375       8.809800   
min        0.000000       6.000000       0.000000       0.000000   
25%        4.800000      31.000000       7.000000      13.000000   
50%        8.400000      39.000000      13.000000      19.000000   
75%       10.600000      48.000000      19.000000      24.000000   
max       14.500000     135.000000     130.000000      87.000000   

         Humidity9am    Humidity3pm   Pressure9am    Pressure3pm  \
count  142806.000000  140953.000000  130395.00000  130432.000000   
mean       68.880831      51.539116    1017.64994    1015.255889   
std        19.029164      20.795902       7.10653       7.037414   
min         0.000000       0.000000     980.50000     977.100000   
25%        57.000000      37.000000    1012.90000    1010.400000   
50%        70.000000      52.000000    1017.60000    1015.200000   
75%        83.000000      66.000000    1022.40000    1020.000000   
max       100.000000     100.000000    1041.00000    1039.600000   

           Cloud9am      Cloud3pm        Temp9am       Temp3pm  
count  89572.000000  86102.000000  143693.000000  141851.00000  
mean       4.447461      4.509930      16.990631      21.68339  
std        2.887159      2.720357       6.488753       6.93665  
min        0.000000      0.000000      -7.200000      -5.40000  
25%        1.000000      2.000000      12.300000      16.60000  
50%        5.000000      5.000000      16.700000      21.10000  
75%        7.000000      7.000000      21.600000      26.40000  
max        9.000000      9.000000      40.200000      46.70000

In [650]:
# check to see the number of rows(entities) and columns (observations) that are available
rainfall_data.shape

(145460, 23)

In [651]:
# Check to see how many columns are non-null and null
rainfall_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145460 entries, 0 to 145459
Data columns (total 23 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   Date           145460 non-null  object 
 1   Location       145460 non-null  object 
 2   MinTemp        143975 non-null  float64
 3   MaxTemp        144199 non-null  float64
 4   Rainfall       142199 non-null  float64
 5   Evaporation    82670 non-null   float64
 6   Sunshine       75625 non-null   float64
 7   WindGustDir    135134 non-null  object 
 8   WindGustSpeed  135197 non-null  float64
 9   WindDir9am     134894 non-null  object 
 10  WindDir3pm     141232 non-null  object 
 11  WindSpeed9am   143693 non-null  float64
 12  WindSpeed3pm   142398 non-null  float64
 13  Humidity9am    142806 non-null  float64
 14  Humidity3pm    140953 non-null  float64
 15  Pressure9am    130395 non-null  float64
 16  Pressure3pm    130432 non-null  float64
 17  Cloud9am       89572 non-null

In [652]:
rainfall_data.columns

Index(['Date', 'Location', 'MinTemp', 'MaxTemp', 'Rainfall', 'Evaporation',
       'Sunshine', 'WindGustDir', 'WindGustSpeed', 'WindDir9am', 'WindDir3pm',
       'WindSpeed9am', 'WindSpeed3pm', 'Humidity9am', 'Humidity3pm',
       'Pressure9am', 'Pressure3pm', 'Cloud9am', 'Cloud3pm', 'Temp9am',
       'Temp3pm', 'RainToday', 'RainTomorrow'],
      dtype='object')

In [653]:
# Optional
# Dropping irrelevant columns from the dataframe using drop method of pandas
# rainfall_data = rainfall_data.drop(['Date','Evaporation','Sunshine', 'Cloud9am','Cloud3pm'])

In [654]:
# Drop all irrelevant columns from my dataset
# Extracting the dependent and independent variables
X = rainfall_data.iloc[:,[1,2,3,4,7,8,9,10,11,12,13,14,15,16,19,20,21]].values
Y = rainfall_data.iloc[:,-1].values

##### The dataframe at the following index [1,2,3,4,7,8,9,10,11,12,13,14,15,16,19,20,21] implies the following column names ['Location', 'MinTemp', 'MaxTemp', 'Rainfall', 'WindGustDir', 'WindGustSpeed', 'WindDir9am', 'WindDir3pm', 'WindSpeed9am', 'WindSpeed3pm', 'Humidity9am', 'Humidity3pm', 'Pressure9am', 'Pressure3pm', 'Temp9am', 'Temp3pm', 'RainToday']

In [655]:
print(X)

[['Albury' 13.4 22.9 ... 16.9 21.8 'No']
 ['Albury' 7.4 25.1 ... 17.2 24.3 'No']
 ['Albury' 12.9 25.7 ... 21.0 23.2 'No']
 ...
 ['Uluru' 5.4 26.9 ... 12.5 26.1 'No']
 ['Uluru' 7.8 27.0 ... 15.1 26.0 'No']
 ['Uluru' 14.9 nan ... 15.0 20.9 'No']]


In [656]:
columns = ['Location', 'MinTemp', 'MaxTemp', 'Rainfall', 'WindGustDir', 'WindGustSpeed', 'WindDir9am', 'WindDir3pm', 'WindSpeed9am', 'WindSpeed3pm', 'Humidity9am', 'Humidity3pm', 'Pressure9am', 'Pressure3pm', 'Temp9am', 'Temp3pm', 'RainToday']
X = pd.DataFrame(data=X, columns = columns, dtype=None, copy=False)
X.head(10)

Location MinTemp MaxTemp Rainfall WindGustDir WindGustSpeed WindDir9am  \
0   Albury    13.4    22.9      0.6           W          44.0          W   
1   Albury     7.4    25.1      0.0         WNW          44.0        NNW   
2   Albury    12.9    25.7      0.0         WSW          46.0          W   
3   Albury     9.2    28.0      0.0          NE          24.0         SE   
4   Albury    17.5    32.3      1.0           W          41.0        ENE   
5   Albury    14.6    29.7      0.2         WNW          56.0          W   
6   Albury    14.3    25.0      0.0           W          50.0         SW   
7   Albury     7.7    26.7      0.0           W          35.0        SSE   
8   Albury     9.7    31.9      0.0         NNW          80.0         SE   
9   Albury    13.1    30.1      1.4           W          28.0          S   

  WindDir3pm WindSpeed9am WindSpeed3pm Humidity9am Humidity3pm Pressure9am  \
0        WNW         20.0         24.0        71.0        22.0      1007.7   
1        WSW          4.0         22.0        44.0        25.0      1010.6   
2        WSW         19.0         26.0        38.0        30.0      1007.6   
3          E         11.0          9.0        45.0        16.0      1017.6   
4         NW          7.0         20.0        82.0        33.0      1010.8   
5          W         19.0         24.0        55.0        23.0      1009.2   
6          W         20.0         24.0        49.0        19.0      1009.6   
7          W          6.0         17.0        48.0        19.0      1013.4   
8         NW          7.0         28.0        42.0         9.0      1008.9   
9        SSE         15.0         11.0        58.0        27.0      1007.0   

  Pressure3pm Temp9am Temp3pm RainToday  
0      1007.1    16.9    21.8        No  
1      1007.8    17.2    24.3        No  
2      1008.7    21.0    23.2        No  
3      1012.8    18.1    26.5        No  
4      1006.0    17.8    29.7        No  
5      1005.4    20.6    28.9        No  
6      1008.2    18.1    24.6        No  
7      1010.1    16.3    25.5        No  
8      1003.6    18.3    30.2        No  
9      1005.7    20.1    28.2       Yes

In [657]:
X.shape

(145460, 17)

In [658]:
print(Y)

['No' 'No' 'No' ... 'No' 'No' nan]


In [659]:
# Visualize all the dependent variables (Y) as a dataframe
pd.DataFrame(data=Y, columns=['RainTomorrow'], dtype=None, copy=True)

RainTomorrow
0                No
1                No
2                No
3                No
4                No
...             ...
145455           No
145456           No
145457           No
145458           No
145459          NaN

[145460 rows x 1 columns]

In [660]:
Y.shape

(145460,)

### Changing the dimension of Y from 1-D to 2-D

In [661]:
Y = Y.reshape(-1,1) #command to change 1-D list to 2-D list 

In [662]:
print(Y)

[['No']
 ['No']
 ['No']
 ...
 ['No']
 ['No']
 [nan]]


# Dealing with Invalid Dataset

In [663]:
from sklearn.impute import SimpleImputer
imputer = SimpleImputer(missing_values=np.nan, strategy='most_frequent') # replaces missing values i.e. nan values with most frequent values.
X = imputer.fit_transform(X)
Y = imputer.fit_transform(Y)

In [664]:
print(Y)

[['No']
 ['No']
 ['No']
 ...
 ['No']
 ['No']
 ['No']]


In [665]:
Y.shape

(145460, 1)

In [666]:
print(X)

[['Albury' 13.4 22.9 ... 16.9 21.8 'No']
 ['Albury' 7.4 25.1 ... 17.2 24.3 'No']
 ['Albury' 12.9 25.7 ... 21.0 23.2 'No']
 ...
 ['Uluru' 5.4 26.9 ... 12.5 26.1 'No']
 ['Uluru' 7.8 27.0 ... 15.1 26.0 'No']
 ['Uluru' 14.9 20.0 ... 15.0 20.9 'No']]


In [667]:
X.shape

(145460, 17)

# Encoding the dataset

> ##### Note all Machine learning models and algorithms only deals with numeric data. Data encoding invloves the transformation of All categorical data into numeric data needed for the Machine learning model.



In [668]:
from sklearn.preprocessing import LabelEncoder
lev1 = LabelEncoder()
X[:,0] = lev1.fit_transform(X[:,0])
lev2 = LabelEncoder()
X[:,4] = lev2.fit_transform(X[:,4])
lev3 = LabelEncoder()
X[:,6] = lev3.fit_transform(X[:,6])
lev4 = LabelEncoder()
X[:,7] = lev4.fit_transform(X[:,7])
lev5 = LabelEncoder()
X[:,-1] = lev5.fit_transform(X[:,-1])
lev6 = LabelEncoder()
Y = lev6.fit_transform(Y)


/usr/local/lib/python3.8/dist-packages/sklearn/preprocessing/_label.py:115: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [669]:
print(X)

[[2 13.4 22.9 ... 16.9 21.8 0]
 [2 7.4 25.1 ... 17.2 24.3 0]
 [2 12.9 25.7 ... 21.0 23.2 0]
 ...
 [41 5.4 26.9 ... 12.5 26.1 0]
 [41 7.8 27.0 ... 15.1 26.0 0]
 [41 14.9 20.0 ... 15.0 20.9 0]]


In [670]:
# Visualize all the independent variables (X) as a dataframe
columns = ['Location', 'MinTemp', 'MaxTemp', 'Rainfall', 'WindGustDir', 'WindGustSpeed', 'WindDir9am', 'WindDir3pm', 'WindSpeed9am', 'WindSpeed3pm', 'Humidity9am', 'Humidity3pm', 'Pressure9am', 'Pressure3pm', 'Temp9am', 'Temp3pm', 'RainToday']
X = pd.DataFrame(data=X, columns= columns, dtype=None, copy=False)
X.head(10)

Location MinTemp MaxTemp Rainfall WindGustDir WindGustSpeed WindDir9am  \
0        2    13.4    22.9      0.6          13          44.0         13   
1        2     7.4    25.1      0.0          14          44.0          6   
2        2    12.9    25.7      0.0          15          46.0         13   
3        2     9.2    28.0      0.0           4          24.0          9   
4        2    17.5    32.3      1.0          13          41.0          1   
5        2    14.6    29.7      0.2          14          56.0         13   
6        2    14.3    25.0      0.0          13          50.0         12   
7        2     7.7    26.7      0.0          13          35.0         10   
8        2     9.7    31.9      0.0           6          80.0          9   
9        2    13.1    30.1      1.4          13          28.0          8   

  WindDir3pm WindSpeed9am WindSpeed3pm Humidity9am Humidity3pm Pressure9am  \
0         14         20.0         24.0        71.0        22.0      1007.7   
1         15          4.0         22.0        44.0        25.0      1010.6   
2         15         19.0         26.0        38.0        30.0      1007.6   
3          0         11.0          9.0        45.0        16.0      1017.6   
4          7          7.0         20.0        82.0        33.0      1010.8   
5         13         19.0         24.0        55.0        23.0      1009.2   
6         13         20.0         24.0        49.0        19.0      1009.6   
7         13          6.0         17.0        48.0        19.0      1013.4   
8          7          7.0         28.0        42.0         9.0      1008.9   
9         10         15.0         11.0        58.0        27.0      1007.0   

  Pressure3pm Temp9am Temp3pm RainToday  
0      1007.1    16.9    21.8         0  
1      1007.8    17.2    24.3         0  
2      1008.7    21.0    23.2         0  
3      1012.8    18.1    26.5         0  
4      1006.0    17.8    29.7         0  
5      1005.4    20.6    28.9         0  
6      1008.2    18.1    24.6         0  
7      1010.1    16.3    25.5         0  
8      1003.6    18.3    30.2         0  
9      1005.7    20.1    28.2         1

In [671]:
# plt.figure(figsize=(8,8))
# sns.scatterplot(x='MaxTemp', y='MinTemp', hue='RainTomorrow')

In [672]:
print(Y)

[0 0 0 ... 0 0 0]


In [673]:
# Visualize all the dependent variables (Y) as a dataframe
pd.DataFrame(data=Y, columns=['RainTomorrow'], dtype=None, copy=True)

RainTomorrow
0                  0
1                  0
2                  0
3                  0
4                  0
...              ...
145455             0
145456             0
145457             0
145458             0
145459             0

[145460 rows x 1 columns]

# Feature Scaling

##### The final step of the machine learning model is feature scaling. It is a method to Standardize the training dataset in a specific range. In feature scaling, all the values are kept in the same range and on the same scale so that no variable dominates the other variable. 

In [674]:
from sklearn.preprocessing import StandardScaler
sc= StandardScaler()
X = sc.fit_transform(X)

In [675]:
print(X)

[[-1.53166617  0.19132753 -0.04135977 ... -0.01407077  0.02310362
  -0.52979545]
 [-1.53166617 -0.75105231  0.26874452 ...  0.03244663  0.387799
  -0.52979545]
 [-1.53166617  0.11279588  0.35331842 ...  0.62166712  0.22733303
  -0.52979545]
 ...
 [ 1.20928479 -1.06517892  0.52246622 ... -0.69632607  0.65037966
  -0.52979545]
 [ 1.20928479 -0.68822699  0.53656187 ... -0.29317521  0.63579185
  -0.52979545]
 [ 1.20928479  0.42692249 -0.45013361 ... -0.30868102 -0.10818671
  -0.52979545]]


In [676]:
# Visualize all the independent variables (X) as a dataframe
columns = ['Location', 'MinTemp', 'MaxTemp', 'Rainfall', 'WindGustDir', 'WindGustSpeed', 'WindDir9am', 'WindDir3pm', 'WindSpeed9am', 'WindSpeed3pm', 'Humidity9am', 'Humidity3pm', 'Pressure9am', 'Pressure3pm', 'Temp9am', 'Temp3pm', 'RainToday']
X = pd.DataFrame(data=X, columns= columns, dtype=None, copy=False)
X.head(10)

Location   MinTemp   MaxTemp  Rainfall  WindGustDir  WindGustSpeed  \
0 -1.531666  0.191328 -0.041360 -0.203581     1.045228       0.327736   
1 -1.531666 -0.751052  0.268745 -0.275097     1.258262       0.327736   
2 -1.531666  0.112796  0.353318 -0.275097     1.471296       0.479465   
3 -1.531666 -0.468338  0.677518 -0.275097    -0.872075      -1.189550   
4 -1.531666  0.835287  1.283631 -0.155903     1.045228       0.100143   
5 -1.531666  0.379803  0.917144 -0.251258     1.258262       1.238108   
6 -1.531666  0.332685  0.254649 -0.275097     1.045228       0.782922   
7 -1.531666 -0.703933  0.494275 -0.275097     1.045228      -0.355043   
8 -1.531666 -0.389807  1.227249 -0.275097    -0.446007       3.058852   
9 -1.531666  0.144209  0.973527 -0.108226     1.045228      -0.886093   

   WindDir9am  WindDir3pm  WindSpeed9am  WindSpeed3pm  Humidity9am  \
0    1.328766    1.366458      0.677819      0.623294     0.081409   
1   -0.221338    1.586813     -1.124341      0.394836    -1.318948   
2    1.328766    1.586813      0.565184      0.851751    -1.630138   
3    0.442992   -1.718521     -0.335896     -1.090136    -1.267083   
4   -1.328556   -0.176032     -0.786436      0.166379     0.651924   
5    1.328766    1.146102      0.565184      0.623294    -0.748432   
6    1.107323    1.146102      0.677819      0.623294    -1.059622   
7    0.664436    1.146102     -0.899071     -0.176307    -1.111487   
8    0.442992   -0.176032     -0.786436      1.080208    -1.422678   
9    0.221549    0.485035      0.114644     -0.861679    -0.592837   

   Humidity3pm  Pressure9am  Pressure3pm   Temp9am   Temp3pm  RainToday  
0    -1.443652    -1.457215    -1.224564 -0.014071  0.023104  -0.529795  
1    -1.297105    -1.026898    -1.119521  0.032447  0.387799  -0.529795  
2    -1.052860    -1.472054    -0.984466  0.621667  0.227333  -0.529795  
3    -1.736746     0.011799    -0.369217  0.171999  0.708731  -0.529795  
4    -0.906314    -0.997221    -1.389630  0.125481  1.175541  -0.529795  
5    -1.394803    -1.234637    -1.479667  0.559644  1.058838  -0.529795  
6    -1.590199    -1.175283    -1.059497  0.171999  0.431562  -0.529795  
7    -1.590199    -0.611419    -0.774381 -0.107106  0.562853  -0.529795  
8    -2.078688    -1.279153    -1.749776  0.203010  1.248480  -0.529795  
9    -1.199407    -1.561085    -1.434649  0.482115  0.956724   1.887521

# Splitting Dataset into Training set and Test set

In [567]:
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X,Y,test_size=0.2,random_state=0)

In [568]:
print(X_train)

[[ 0.22535368  1.03946939  0.07140543 ...  0.68369032  0.08145488
  -0.52979545]
 [ 1.42012717 -0.45263203  0.11369237 ... -0.41722163  0.22733303
  -0.52979545]
 [ 0.50647685 -0.20133073 -0.14002932 ... -0.06058818 -0.02065982
   1.88752093]
 ...
 [ 1.0687232   0.75675544  0.93124006 ...  1.10234698  1.07342629
  -0.52979545]
 [ 0.57675765 -0.04426743 -0.16822062 ...  0.01694083 -0.28324049
   1.88752093]
 [ 1.63096955 -0.0285611  -0.91529006 ... -0.35519842 -0.76463838
  -0.52979545]]


In [569]:
print(Y_train)

[1 0 0 ... 0 0 0]


# Training the Model

> Importing the measurement metrics (e.g accuracy score, confusion matrix and classification report)

In [570]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

> By definition a confusion matrix C is such that C_{i, j}
is equal to the number of observations known to be in group i and predicted to be in group j.

> Thus in binary classification, the count of true negatives is
C_{0,0}, false negatives is C_{1,0}, true positives is
C_{1,1} and false positives is C_{0,1}.

### 1. Logistic Regression


In [594]:
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression()
lr.fit(X_train, Y_train)
y_predict= lr.predict(X_test)

# Confusion Matrix for the Logistic Regression
print('Confusion Matrix (Logistic Regression):')
print(confusion_matrix(Y_test,y_predict))
print("\n*************************************************************\n")

# Classification Report for the Logistic Regression
print('Classification Report (Logistic Regression):\n' + classification_report(Y_test,y_predict))
print("*************************************************************\n")

# Accuracy Score for the Logistic Regression
print('Accuracy Score (Logistc Regression):\n')
print(accuracy_score(Y_test,y_predict))


Confusion Matrix (Logistic Regression):
[[21603  1123]
 [ 3488  2878]]

*************************************************************

Classification Report (Logistic Regression):
              precision    recall  f1-score   support

           0       0.86      0.95      0.90     22726
           1       0.72      0.45      0.56      6366

    accuracy                           0.84     29092
   macro avg       0.79      0.70      0.73     29092
weighted avg       0.83      0.84      0.83     29092

*************************************************************

Accuracy Score (Logistc Regression):

0.8415028186443009


### 2. Decision Tree

In [596]:
from sklearn.tree import DecisionTreeClassifier
dt = DecisionTreeClassifier()
dt.fit(X_train, Y_train)
y_predict= dt.predict(X_test)

# Confusion Matrix for the Decision Tree Classifier
print('Confusion Matrix (Decision Tree Classifier):')
print(confusion_matrix(Y_test,y_predict))
print("\n*************************************************************\n")

# Classification Report for the Decision Tree
print('Classification Report (Decision Tree):\n' + classification_report(Y_test,y_predict))
print("*************************************************************\n")

# Accuracy Score for the Decision Tree
print('Accuracy Score (Decision Tree):\n')
print(accuracy_score(Y_test,y_predict))

Confusion Matrix (Decision Tree Classifier):
[[19462  3264]
 [ 3087  3279]]

*************************************************************

Classification Report (Decision Tree):
              precision    recall  f1-score   support

           0       0.86      0.86      0.86     22726
           1       0.50      0.52      0.51      6366

    accuracy                           0.78     29092
   macro avg       0.68      0.69      0.68     29092
weighted avg       0.78      0.78      0.78     29092

*************************************************************

Accuracy Score (Decision Tree):

0.7816925615289426


### 3. Random Forest Classifier





In [588]:
from sklearn.ensemble import RandomForestClassifier
cl = RandomForestClassifier(n_estimators =150, random_state=0)
cl.fit(X_train,Y_train)
y_predict= cl.predict(X_test)

# Confusion Matrix for Random Forest
print('Confusion Matrix (Random Forest):')
print(confusion_matrix(Y_test,y_predict))
print("\n*************************************************************\n")

# Classification Report for the Random Forest
print('Classification Report (Random Forest):\n' + classification_report(Y_test,y_predict))
print("*************************************************************\n")

# Accuracy Score for the Random Forest
print('Accuracy Score (Random Forest):\n')
print(accuracy_score(Y_test,y_predict))


Confusion Matrix (Random Forest):
[[21697  1029]
 [ 3269  3097]]

*************************************************************

Classification Report (Random Forest):
              precision    recall  f1-score   support

           0       0.87      0.95      0.91     22726
           1       0.75      0.49      0.59      6366

    accuracy                           0.85     29092
   macro avg       0.81      0.72      0.75     29092
weighted avg       0.84      0.85      0.84     29092

*************************************************************

Accuracy Score (Random Forest):

0.8522617901828682


> From our confusion matrix above, there are 21697 True positive (TP) values and 1029 False Positive (FP) values

# 4. XGBoost Classifier

In [597]:
import xgboost as xgb
xgb = xgb.XGBClassifier()
xgb.fit(X_train,Y_train)
y_predict= xgb.predict(X_test)

# Confusion Matrix for XGBoost Classifier
print('Confusion Matrix (XGBoost Classifier):')
print(confusion_matrix(Y_test,y_predict))
print("\n*************************************************************\n")

# Classification Report for the XGBoost Classifier
print('Classification Report (XGBoost Classifier):\n' + classification_report(Y_test,y_predict))
print("*************************************************************\n")

# Accuracy Score for the XGBoost Classifier
print('Accuracy Score (XGBoost Classifier):\n')
print(accuracy_score(Y_test,y_predict))


Confusion Matrix (XGBoost Classifier):
[[21719  1007]
 [ 3418  2948]]

*************************************************************

Classification Report (XGBoost Classifier):
              precision    recall  f1-score   support

           0       0.86      0.96      0.91     22726
           1       0.75      0.46      0.57      6366

    accuracy                           0.85     29092
   macro avg       0.80      0.71      0.74     29092
weighted avg       0.84      0.85      0.83     29092

*************************************************************

Accuracy Score (XGBoost Classifier):

0.8478963288876668


# Decoding the dataset of the dependent variable Y

In [601]:
# convert the predictiion results to Yes or No
y_predict = lev6.inverse_transform(y_predict)

In [602]:
print(y_predict)

['No' 'No' 'No' ... 'No' 'No' 'No']


In [603]:
print(Y_test)

[1 1 0 ... 1 0 0]


In [604]:
Y_test = lev6.inverse_transform(Y_test)

In [605]:
# Print the Y_test as a 1-D array
print(Y_test)

['Yes' 'Yes' 'No' ... 'Yes' 'No' 'No']


In [606]:
# Y_test and y_predict are 1-D arrays. Convert Y_test and y_predict to 2-D array
Y_test = Y_test.reshape(-1,1)
y_predict = y_predict.reshape(-1,1)

### Concatenate the Y_test and y_predict into a single dataframe

In [614]:
df = np.concatenate((Y_test,y_predict), axis= 1)
dataframe = pd.DataFrame(df, columns=['Rain Tommorrow', 'Prediction of Rain'])
# print(df)
print(dataframe)

      Rain Tommorrow Prediction of Rain
0                Yes                 No
1                Yes                 No
2                 No                 No
3                 No                Yes
4                 No                 No
...              ...                ...
29087             No                 No
29088             No                 No
29089            Yes                 No
29090             No                 No
29091             No                 No

[29092 rows x 2 columns]


In [616]:
dataframe.to_csv('prediction.csv')

# Checking Model Accuracy

In [600]:
from sklearn.metrics import accuracy_score
accuracy_score(Y_test,y_predict)

0.8522617901828682